File: 백준/Gold/4179. 불！/불！.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [R, C] = input[0].split(" ").map(Number);
const matrix = [];

// 매트릭스 초기화
for (let i = 1; i <= R; i++) {
  let line = input[i].split("");
  matrix.push(line);
}

let jihunQ = [];
let fireQ = [];

// 초기 위치 설정
for (let i = 0; i < R; i++) {
  for (let j = 0; j < C; j++) {
    if (matrix[i][j] === "J") {
      jihunQ.push([i, j, 0]);
      matrix[i][j] = "."; // 지훈이 시작점을 빈 공간으로
    } else if (matrix[i][j] === "F") {
      fireQ.push([i, j]);
    }
  }
}

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

function bfs() {
  while (jihunQ.length > 0) {

    let fireSize = fireQ.length;
    for (let f = 0; f < fireSize; f++) {
      let [fX, fY] = fireQ.shift();

      for (let i = 0; i < 4; i++) {
        let nx = fX + dx[i];
        let ny = fY + dy[i];

        if (nx >= 0 && nx < R && ny >= 0 && ny < C && matrix[nx][ny] === ".") {
          matrix[nx][ny] = "F";
          fireQ.push([nx, ny]);
        }
      }
    }


    let jihunSize = jihunQ.length;
    for (let j = 0; j < jihunSize; j++) {
      let [jX, jY, time] = jihunQ.shift();


      if (jX === 0 || jX === R - 1 || jY === 0 || jY === C - 1) {
        return time + 1;
      }

      for (let i = 0; i < 4; i++) {
        let nx = jX + dx[i];
        let ny = jY + dy[i];

        if (nx >= 0 && nx < R && ny >= 0 && ny < C && matrix[nx][ny] === ".") {
          matrix[nx][ny] = "J";
          jihunQ.push([nx, ny, time + 1]);
        }
      }
    }
  }

  return -1; // 탈출 불가능
}

const result = bfs();
console.log(result !== -1 ? result : 'IMPOSSIBLE');

File: 백준/Gold/4179. 불！/불！.py
from collections import deque

R, C = map(int, input().split())
jihun_q = deque()
fire_q = deque()
matrix = []

for i in range(R):
    matrix.append(list(input()))

def bfs():
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    while jihun_q:
        # 1. 불 먼저 퍼뜨리기
        fire_size = len(fire_q)
        for _ in range(fire_size):
            x, y = fire_q.popleft()
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < R and 0 <= ny < C and matrix[nx][ny] == ".":
                    matrix[nx][ny] = "F"
                    fire_q.append([nx, ny])
        
        # 2. 지훈이 이동
        jihun_size = len(jihun_q)
        for _ in range(jihun_size):
            x, y, time = jihun_q.popleft()
            
            # 경계에 도달하면 탈출!
            if x == 0 or x == R-1 or y == 0 or y == C-1:
                return time + 1
            
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < R and 0 <= ny < C and matrix[nx][ny] == ".":
                    matrix[nx][ny] = "J"  # 방문 표시
                    jihun_q.append([nx, ny, time + 1])
    
    return -1  # 탈출 불가능

# 초기 위치 설정
for i in range(R):
    for j in range(C):
        if matrix[i][j] == "J":
            jihun_q.append([i, j, 0])
            matrix[i][j] = "."  # 지훈이 시작점을 일반 통로로 변경
        elif matrix[i][j] == "F":
            fire_q.append([i, j])

result = bfs()

if result != -1:
    print(result)
else:
    print('IMPOSSIBLE')